### PR TITLE
Allow for registration of DRep scripts

### LINKED ISSUE #559: [FEATURE] -  Add support for registering scripts as DReps.
## What
Add support for registering scripts as DReps. 

Expand `cardano-cli conway governance drep registration-certificate` so that it can also take `--script-file` and `--script-hash` as inputs. 

The new flags should be alternatives to the existing --drep-verification-key, --drep-verification-key-file, and --drep-key-hash  

The resulting certificate must comply with the conway cddl, where: 

```
reg_drep_cert = (16, drep_credential, coin, anchor / null)

drep_credential = credential

credential =
  [  0, addr_keyhash
  // 1, scripthash
  ]
```

## Why
From CIP1694

> Registered DReps are identified by a credential that can be either:
> 
> A verification key (Ed25519)
> A native or Plutus script

## Personas

- [ ] SPOs
- [ ] dApp Devs
- [ ] Exchanges
- [ ] Wallets
- [ ] 3rd party tools
- [x] ADA holders 
- [x] DReps

## Definition of Done (DoD)

- [ ] [Acceptance Criteria + User Stories & DoD](https://input-output.atlassian.net/wiki/spaces/QA/pages/3584229468/Acceptance+Criteria+DoD) created and singed-off (by PO, dev & test owners)
- [ ] Builds successfully on CI
- [ ] Code & Test review (as per Acceptance Criteria)
- [ ] There is documentation and/or examples for the new functionality (usage/response)
- [ ] Log/record changes on **Vnext** (or similar depending on what we adopt)
- [ ] Ticket number(s) included in PR description
- [ ] All Acceptance Criteria met and covered by dev/unit/property/integration tests
- [ ] System/E2E automated tests + System Test Engineer Owner Sign-off

> **_NOTE:_**  Ideally, we should merge only fully implemented and tested features into the master branch. 
> So all the above steps are required for the PR to be merged.  
> In order to avoid the PRs becoming stale and requiring to be rebased on master, these can be merged 
> after a reasonable time (current agreement is 3 days) if the System Test Engineer Owner's sign-off 
> was not provided (last step in the DoD).

> **_IMPORTANT:_** Any deviation from the plan should be discussed and agreed as a comment in the Feature file. 

## Sign-off

- [ ] Product Owner
- [ ] Dev Owner
- [ ] System Test Engineer Owner

# Related PRs

1.  PR # here

### PR DESCRIPTION
**Pull request by carlhammann**
# Changelog

```yaml
- description: |
     - add support for registration of script DReps;
     - rework the `conway governance hash` command: separate sub-commands for anchor data and scripts
# uncomment types applicable to the change:
  type:
  # - feature        # introduces a new feature
  - breaking       # the API has changed in a breaking way
  # - compatible     # the API has changed but is non-breaking
  # - optimisation   # measurable performance improvements
  - improvement    # QoL changes e.g. refactoring
  - bugfix         # fixes a defect
  # - test           # fixes/modifies tests
  # - maintenance    # not directly related to the code
  # - release        # related to a new release preparation
  # - documentation  # change in code docs, haddocks...
```

# Context

This PR closes #559. The issue proposes adding a `--script-hash` and a `--script-file` flag to the `conway governance drep registration-certificate` command. In the spirit of #442, I choose to implement only one of these flags, as `--drep-script-hash`, and instead add a new sub-command to the `conway governance hash` command, to enable computation of script hashes. This is done in [bf65a3a](https://github.com/IntersectMBO/cardano-cli/pull/563/commits/bf65a3a337d81b8bb1eca251d1d826e89616a202).

# How to trust this PR

Highlight important bits of the PR that will make the review faster. If there are commands the reviewer can run to observe the new behavior, describe them.

# Checklist

- [ ] Commit sequence broadly makes sense and commits have useful messages
- [ ] New tests are added if needed and existing tests are updated. See [Running tests](https://github.com/input-output-hk/cardano-node-wiki/wiki/Running-tests) for more details
- [ ] Self-reviewed the diff

### CODE REVIEW

File: cardano-cli/src/Cardano/CLI/EraBased/Options/Governance/Hash.hs [hunk 1]
```diff
@@ -3,48 +3,74 @@
...
     $ Opt.info
-        ( fmap Cmd.GovernanceHashCmd
-            (GovernanceHashCmdArgs eon
-               <$> pGovernanceHashSource
+        ( fmap Cmd.GovernanceHashAnchorDataCmd
+            (Cmd.GovernanceHashAnchorDataCmdArgs eon
+               <$> pGovernanceAnchorDataHashSource
                <*> optional pOutputFile))
-    $ Opt.progDesc "Compute the hash to pass to the various --*-hash arguments of governance commands."
+    $ Opt.progDesc "Compute the hash of some anchor data."
```
Comment (smelc) [inline cardano-cli/src/Cardano/CLI/EraBased/Options/Governance/Hash.hs]:
```suggestion
    $ Opt.progDesc "Compute the hash of some anchor data (to then pass it to other governance commands)."
```

Optional

Rationale: be a bit more helpful to users.

File: cardano-cli/src/Cardano/CLI/EraBased/Options/Governance/Hash.hs [hunk 2]
```diff
@@ -3,48 +3,74 @@
...
+pGovernanceHashScriptCmd era = do
+  eon <- forEraMaybeEon era
+  return
+    $ subParser "script"
+    $ Opt.info
+        ( fmap Cmd.GovernanceHashScriptCmd
+            (Cmd.GovernanceHashScriptCmdArgs eon
+               <$> pScript
+               <*> optional pOutputFile))
+    $ Opt.progDesc "Compute the hash of a script."
```
Comment (smelc) [inline cardano-cli/src/Cardano/CLI/EraBased/Options/Governance/Hash.hs]:
```suggestion
    $ Opt.progDesc "Compute the hash of a script (to then pass it to other governance commands)."
```

Optional again.

File: cardano-cli/test/cardano-cli-golden/Test/Golden/Governance/DRep.hs [hunk 3]
```diff
@@ -8,13 +8,10 @@ module Test.Golden.Governance.DRep where
 #endif
 
 import           Control.Monad
-
-#ifdef UNIX
```
Comment (smelc) [inline cardano-cli/test/cardano-cli-golden/Test/Golden/Governance/DRep.hs]: Is this change intended?

**Comment by smelc** [comment]
Nice design :heart:

File: cardano-cli/test/cardano-cli-golden/Test/Golden/Governance/DRep.hs (hunk 3 above)
Comment (carlhammann) [reply cardano-cli/test/cardano-cli-golden/Test/Golden/Governance/DRep.hs]: No, again... Seems like an issue with the formatter?

File: cardano-cli/src/Cardano/CLI/EraBased/Commands/Governance/Hash.hs [hunk 4]
```diff
@@ -16,21 +17,31 @@ import           Cardano.CLI.Types.Common
...
+    , toHash  :: !GovernanceAnchorDataHashSource
     , moutFile :: !(Maybe (File () Out)) -- ^ The output file to which the hash is written
   } deriving Show
 
-data GovernanceHashSource
-  = GovernanceHashSourceBinaryFile (File ProposalText In)
-  | GovernanceHashSourceTextFile (File ProposalText In)
-  | GovernanceHashSourceText Text
+data GovernanceAnchorDataHashSource
+  = GovernanceAnchorDataHashSourceBinaryFile (File ProposalText In)
```
Comment (carbolymer) [inline cardano-cli/src/Cardano/CLI/EraBased/Commands/Governance/Hash.hs:32]: `ProposalText` for binary files is quite misleading here. We should use use a different type here. Not necessarily within the scope of this PR.

**Review by carbolymer — APPROVED** [review]

File: cardano-cli/src/Cardano/CLI/EraBased/Commands/Governance/Hash.hs (hunk 4 above)
Comment (carlhammann) [reply cardano-cli/src/Cardano/CLI/EraBased/Commands/Governance/Hash.hs:32]: I opened an issue for that: #568
